pkg/controller/remotemachineset/remotemachineset_controller.go

```diff
@@ -0,0 +1,502 @@
+	// Load the kubeconfig secret for communicating with the remote cluster:
+
+	// Step 1: Get the secret ref from the cluster
+	//
+	// TODO: The cluster controller should watch secrets until it
+	// sees a kubeconfig secret named after the cluster and then
+	// update cluster status with a reference to that
+	// secret. Until then we'll look for a
+	// "clustername-kubeconfig" secret.
+	//
+	// clusterConfig := cluster.Status.AdminKubeconfig
+	// if clusterConfig == nil {
+	// 	return nil, fmt.Errorf("unable to retrieve AdminKubeConfig from cluster status")
+	// }
+	secretName := clusterDeployment.Name + "-kubeconfig"
+
+	// Step 2: Retrieve secret
+	secret, err := c.kubeClient.CoreV1().Secrets(clusterDeployment.Namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	secretData := secret.Data["kubeconfig"]
+
+	// Step 3: Generate config from secret data
+	config, err := clientcmd.Load(secretData)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{})
+	restConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Step 4: Return client for remote cluster
+	return clusterapiclient.NewForConfig(restConfig)
+}
+
+// syncClusterDeployment will sync the clusterdeployment with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncClusterDeployment(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing clusterdeployment")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing clusterdeployment")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	clusterDeployment, err := c.clusterDeploymentLister.ClusterDeployments(namespace).Get(name)
+	if err != nil {
+		c.logger.Errorf("Error fetching clusterdeployment %v/%v", namespace, name)
+		return err
+	}
+
+	clusterapiCluster, err := c.clusterInformer.Clusters(clusterDeployment.Namespace).Get(clusterDeployment.Name)
+	if err != nil {
+		return fmt.Errorf("error retrieving cluster object: %v", err)
+	}
+
+	clusterDeploymentStatus, err := controller.ClusterStatusFromClusterAPI(clusterapiCluster)
+	if err != nil {
+		return fmt.Errorf("error fetching cluster status: %v", err)
+	}
+
+	if !clusterDeploymentStatus.ClusterAPIInstalled {
+		c.logger.Debugf(errorMsgClusterAPINotInstalled)
+		return nil
+	}
+
+	remoteClusterAPIClient, err := c.buildRemoteClient(clusterDeployment)
+	if err != nil {
+		return fmt.Errorf("error bulding remoteclusterclient connection: %v", err)
+	}
+
+	err = c.syncMachineSets(clusterDeployment, remoteClusterAPIClient)
+	if err != nil {
+		return err
+	}
+
+	err = c.syncCluster(clusterapiCluster, clusterDeployment, remoteClusterAPIClient)
```

Comment:
I don't think it makes much of a difference right now, but I would sync the cluster before syncing the machinesets